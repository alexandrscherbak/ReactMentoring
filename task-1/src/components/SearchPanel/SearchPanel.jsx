import React from 'react';
import Logo from '../Logo/Logo.jsx';
import {connect} from 'react-redux';
import SearchValueActions from '../../redux/actions/search-value.action.js';
import SearchByActions from '../../redux/actions/search-by.action.js';

class _SearchPanel extends React.Component {
	constructor(props) {
		super(props);
	}

	searchClicked() {
		this.props.searchClicked();
	}

	render() {
		return (
			<div className="search-panel-container">
				<Logo />
				<div className="search-panel">
					<h1>Find your movie</h1>
					<input value={this.props.searchValue} onChange={evt => this.props.updateSearchValue(evt.target.value)}/>
					<div className="search-buttons-container">
						<div className="search-by-buttons-container">
							<span>Search By</span>
							<button
								onClick={() => this.props.updateSearchBy('title')}
								className={this.props.searchBy === 'title' ? 'active-button' : ''}
							>Title</button>
							<button
								onClick={() => this.props.updateSearchBy('genres')}
								className={this.props.searchBy === 'genres' ? 'active-button' : ''}
							>Genre</button>
						</div>
						<div>
							<button
								onClick={() => this.searchClicked()}
							>Search</button>
						</div>
					</div>
				</div>
				<style jsx>
					{`
						.search-panel-container {
							background-image: url('headerBackground.jpeg');
							padding: 20px 100px;
						}
						.search-panel {
							color: #b9b9b9;
						}
						h1, span {
							text-transform: uppercase;
						}
						.search-buttons-container {
							display: flex;
							justify-content: space-between;
						}
						.search-by-buttons-container {
							display: flex;
						}
						input {
							width: 50%;
							margin-bottom: 20px;
						}
						button.active-button {
							background-color: pink;
						}
					`}
				</style> 
			</div>
		);
	}
};

const mapStateToProps = state => {
	return {
		searchValue: state.searchValue,
		searchBy: state.searchBy,
	};
}

const mapDispatchToProps = dispatch => {
	return {
		updateSearchValue: searchValue => dispatch(SearchValueActions.setSearchValue(searchValue)),
		updateSearchBy: searchBy => dispatch(SearchByActions.setSearchBy(searchBy)),
	};
}

const SearchPanel = connect(
	mapStateToProps,
	mapDispatchToProps,
)(_SearchPanel);

export default SearchPanel;
