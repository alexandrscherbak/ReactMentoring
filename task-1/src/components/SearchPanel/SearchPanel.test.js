import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { SearchPanel, mapStateToProps } from './SearchPanel.jsx';
import renderer from 'react-test-renderer';

Enzyme.configure({ adapter: new Adapter() });

const props = {
	searchValue: '123',
	searchBy: 'title',
	searchClicked: jest.fn(),
	updateSearchValue: jest.fn(),
	updateSearchBy: jest.fn(),
};

function setup() {
	return mount(<SearchPanel {...props} />);
}

describe('SearchPanel', () => {
	test('snapshot', () => {
		const component = renderer.create(
			<SearchPanel {...props}/>,
		);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('should render self and subcomponents', () => {
		const enzymeWrapper = setup();
		expect(enzymeWrapper.find('input').props().value).toBe(props.searchValue);
		expect(enzymeWrapper.find('.active-button').text()).toBe('Title');
	});
	it('should call searchClicked if button clicked', () => {
		const enzymeWrapper = setup();
		const button = enzymeWrapper.find('.search-button');
		expect(props.searchClicked.mock.calls.length).toBe(0);
		button.props().onClick();
		expect(props.searchClicked.mock.calls.length).toBe(1);
	});
	it('should call updateSearchValue if input changed', () => {
		const enzymeWrapper = setup();
		const input = enzymeWrapper.find('input');
		expect(props.updateSearchValue.mock.calls.length).toBe(0);
		input.props().onChange({target: {value: 123}});
		expect(props.updateSearchValue.mock.calls.length).toBe(1);
	});
	it('should call updateSearchBy if button clicked', () => {
		const enzymeWrapper = setup();
		const button1 = enzymeWrapper.find('.search-by-buttons-container').childAt(1);
		const button2 = enzymeWrapper.find('.search-by-buttons-container').childAt(2);
		expect(props.updateSearchBy.mock.calls.length).toBe(0);
		button1.props().onClick();
		expect(props.updateSearchBy.mock.calls.length).toBe(1);
		button2.props().onClick();
		expect(props.updateSearchBy.mock.calls.length).toBe(2);
	});
	it('should change active button if props.searchBy changed', () => {
		const enzymeWrapper = setup();
		expect(enzymeWrapper.find('.active-button').text()).toBe('Title');
		enzymeWrapper.setProps({searchBy: 'genres'});
		expect(enzymeWrapper.find('.active-button').text()).toBe('Genre');
	});
	it('mapStateToProps should return props from state', () => {
		const state = {
			searchValue: 'searchValue',
			searchBy: 'searchBy',
		};
		expect(mapStateToProps(state)).toEqual({
			searchValue: 'searchValue',
			searchBy: 'searchBy',
		});
	});
});
