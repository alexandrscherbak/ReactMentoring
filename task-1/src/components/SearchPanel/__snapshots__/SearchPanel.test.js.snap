// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPanel snapshot 1`] = `
<div
  className="jsx-1049508457 search-panel-container"
>
  <div
    className="jsx-2852891300"
  >
    <span
      className="jsx-2852891300"
    >
      netflixroulette
    </span>
  </div>
  <div
    className="jsx-1049508457 search-panel"
  >
    <h1
      className="jsx-1049508457"
    >
      Find your movie
    </h1>
    <input
      className="jsx-1049508457"
      onChange={[Function]}
      value="123"
    />
    <div
      className="jsx-1049508457 search-buttons-container"
    >
      <div
        className="jsx-1049508457 search-by-buttons-container"
      >
        <span
          className="jsx-1049508457"
        >
          Search By
        </span>
        <button
          className="jsx-1049508457 active-button"
          onClick={[Function]}
        >
          Title
        </button>
        <button
          className="jsx-1049508457 "
          onClick={[Function]}
        >
          Genre
        </button>
      </div>
      <div
        className="jsx-1049508457"
      >
        <button
          className="jsx-1049508457 search-button"
          onClick={[Function]}
        >
          Search
        </button>
      </div>
    </div>
  </div>
</div>
`;
